design/nfs/nfs-quota.md

```diff
@@ -0,0 +1,120 @@
+metadata:
+  name: standard-xfs
+parameters:
+  fsType: xfs
+mountOptions:
+  - prjquota
+...
+```
+
+Then the underlying PersistentVolumeClaim should be used that StorageClass 
+
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: nfs-default-claim
+spec:
+  storageClassName: "standard-xfs"
+  accessModes:
+  - ReadWriteOnce
+...
+```
+
+If the above conditions are met then the Rook NFS Provisioner will create projects and set the quota limit using [xfs_quota](https://linux.die.net/man/8/xfs_quota) before creating PersistentVolumes based on `.spec.resources.requests.storage` field in PersistentVolumeClaim. Otherwise the Rook NFS Provisioner will provision a PersistentVolumes without creating setting the quota.
+
+To creating the project, Rook NFS Provisioner will invoke the following command
+
+> xfs_quota -x -c project -s -p '*nfs_pv_directory* *project_id*' *projects_file*
+
+And setting quota with the command
+
+> xfs_quota -x -c 'limit -p bhard=*size* *project_id*' *projects_file*
+
+which
+
+1. *nfs_pv_directory* is sub-directory from exported directory that used for NFS PV.
+1. *project_id* is unique id `uint16` 1 to 65535.
+1. *size* is size of quota as requested.
+1. *projects_file* is file that contains *project quota block* for persisting quota state purpose. In case the Rook NFS Provisioner pod is killed, Rook NFS Provisioner pod will restore the quota state based on *project quota block* entries in *projects_file* at startup.
+1. *project quota block* is combine of *project_id*:*nfs_pv_directory*:*size*
+
+Since Rook NFS has possibility to create more than one of nfs share/export that have different underlying volume directory, the *projects_file* will saved on each underlying volume directory. So each nfs share/export will have different *projects_file* and *project_file* will still persist. The *projects_file* will ony created if underlying volume directory mounted as xfs with prjquota mount options. This mean the existence of *project_file* will indicate if quota was enabled. The hierarchy of directory will look like
+
+```console
+/
+├── underlying-volume-A (export A) (mounted as xfs with prjquota mount options)
+│   ├── projects_file
+│   ├── nfs-pv-a (PV-A) (which quota created for)
+│   │   ├── data (from PV-A)
+│   └── nfs-pv-b (PV-B) (which quota created for)
+│       └── data (from PV-B)
+├── underlying-volume-B (export B) (mounted as xfs with prjquota mount options)
+│   ├── projects_file
+│   └── nfs-pv-c (PV-C) (which quota created for)
+└── underlying-volume-C (export C) (not mounted as xfs)
+    └── nfs-pv-d (PV-D) (quota not created)
+```
+
+The hierarchy above is example Rook NFS has 3 nfs share/exports (A, B and C). *project_file* inside underlying-volume-A will contains *project quota block* like
+
+```
+1:/underlying-volume-A/nfs-pv-a:size
+2:/underlying-volume-A/nfs-pv-b:size
+```
+
+*project_file* inside underlying-volume-B will look like
+
+```
+3:/underlying-volume-B/nfs-pv-c:size
```

Comment:
does this need to be 3 (unique and not conflicting with the previous 2 projects?  This is on a separate underlying volume, so we could reuse 1 or 2 here right? I'm hoping we don't have to manage a unique set of project IDs across underlying volumes

Reply:
Yes, you are right. We should have a different set of project IDs across underlying volumes